### ISSUE TITLE
Unable to Create Contacts "Property properties doesn't exist".

### ISSUE THREAD

**Issue by axyl** [hubspot-net/HubSpot.NET#50]
This is for the v1-preview branch...

When trying to create (or CreateorUpdate) I'm getting an error back from the REST response saying `"Property \"properties\" does not exist","error":"PROPERTY_DOESNT_EXIST","name":"properties"}],"status":"error","message":"Property values were not valid"...`

It looks like the properties data member is being included for the Contact when sending off the JSON.

**Comment by Psypher9** [comment]
Hey @axyl! Thanks for bringing this up. Would you be able to post a safe copy of the payload you're sending? Thank you!

**Comment by axyl** [comment]
Sample call...
`			var contact = api.Contact.Create(new ContactHubSpotModel()
			{
				Email = "john@gmail.com",
				FirstName = "Terry",
				LastName = "Terrence",
				Address= "10 Something Street",
				City= "Heresville",
				ZipCode= "2022",
				State= "NSW",
				Phone = "02 8832 2213",
			});`

Which produces... (in the NewtonsoftResetSharpSerializer) 

`{"properties":[{"property":"email","value":"john@gmail.com"},{"property":"firstname","value":"Terry"},{"property":"lastname","value":"Terrence"},{"property":"phone","value":"02 8832 2213"},{"property":"address","value":"10 Something Street"},{"property":"city","value":"Heresville"},{"property":"state","value":"NSW"},{"property":"zip","value":"2022"},_{"property":"properties","value":"System.Collections.Generic.Dictionary`2[System.String,HubSpot.NET.Api.Contact.Dto.ContactProperty]"}_]}`

Removing the DataMember from the ContractHubSpotModel.Properties property then obviously removes that property from the generated json....which then works.  I haven't applied any brainpower to working out why you might be trying to include that in the payload.

**Comment by Psypher9** [comment]
Thank you for posting that! For the Create, Update and CreateOrUpdate endpoints the POST body is supposed to only contain an array of property value pairs, [see here for more details](https://developers.hubspot.com/docs/methods/contacts/update_contact). 

However since it looks like we're complying with the API spec, I'd need to look closer at this to see why that might be happening.

**Comment by Psypher9** [comment]
Snap. I just saw what you're talking about! 

Sorry. I know exactly what caused that. I've been working on centralizing the 'NameValuePair' , 'PropertyValuePair' collection creation and I probably introduced that because I'm relying on reflection.

Sorry again for my initial misunderstanding. I've been reading this from my phone.  Thanks for catching this!

**Comment by MischaWeerwag** [comment]
Any updates? It looks like ToPropertyTransportModel is not catching the 'properties' dictionary. 

I have 'quick fixed' this like this but there are better solutions for sure:

`                else if (prop.PropertyType.IsGenericType && value is IDictionary<string, ContactProperty> pairs1)
                {
                    foreach (var item in pairs1)
                    {
                        Properties.Add(new PropertyValuePair(item.Key, item.Value.Value.ToString()));
                    }
                    continue;
                }`

**Comment by ricktiger** [comment]
I am on HubSpot.NET 0.8.13.0 doing Contact Create and I am getting this...
Error: HubSpot.NET.Core.HubSpotException: Error from HubSpot, 
Response = {"validationResults":[{"isValid":false,"message":"Property \"properties\" does not exist","error":"PROPERTY_DOESNT_EXIST","name":"properties"}],"status":"error","message":"Property values were not valid","correlationId":"480e8d32-775a-47c7-94f7-b957d856b8f3","requestId":"16aaf6312a430bcf237b108ccea8c468"}

Please advise how to get around the error or what version to switch to that works!

ContactHubSpotModel contact = HsAPI.Contact.Create(new ContactHubSpotModel()
{
    Email = bcCart.Email,
    FirstName = bcCustomer.FirstName,
    LastName = bcCustomer.LastName,
    Phone = bcCustomer.Phone,
    Company = "x"
});

**Comment by Psypher9** [comment]
@ricktiger Are you using 0.8.1.3 or building from the v1-preview branch?

**Comment by Psypher9** [comment]
If I get a moment this week, I will try to look back at this. Sorry that you're still having issues @ricktiger. I know where the issue is, I've just not had a moment to patch it yet.

**Comment by ricktiger** [comment]
We downloaded 0.6.17 source code and added it to our Solution and that version seems to be working for us.  Thank you.

 

RickTiger
Rick Pence

**Comment by ricktiger** [comment]
Thanks, I got past that issue.

 

Any chance you could please help with this question?

 

How would I connect deal with contact using the HubSpot.NET wrapper?  

Right now my Deal.Create code using HubSpot.NET looks like this...


DealHubSpotModel deal = HsAPI.Deal.Create(new DealHubSpotModel()
{
Name = arHdr[i].ArInvoiceHistoryDetails[d].ItemCodeDesc,
Sku = arHdr[i].ArInvoiceHistoryDetails[d].ItemCode,
Amount = Convert.ToDouble(arHdr[i].ArInvoiceHistoryDetails[d].ExtensionAmt),
ProductId = arHdr[i].ArInvoiceHistoryDetails[d].ItemCode
});

 

And I want to  add the Contact Vid so it links to the HsAPI.Deal.Create.

Thanks Much!

Rick Pence



-- 
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

**Comment by odowdbrendan** [comment]
`var contact = api.Contact.Create(new ContactHubSpotModel()
                {
                    Email = emailAddress,
                    FirstName = firstName,
                    LastName = lastName,
                    Phone = phone                    
                });`

Is it only possible to add the standard properties using this wrapper? I have some custom properties that I would need to add

**Comment by ricktiger** [comment]
Yes, you can add custom properties in HubSpot, but you also have to add them to the appropriate model files in HubSpot.NET.  Then code for them.  Property names with spaces become property names with underscores.  Good luck!

**Comment by ricktiger** [comment]
Add your new custom properties that you added via HubSpot to this file ContactHubSpotModel.cs. 
Property names with spaces become property names with underscores.

**Comment by JesseKlaasse** [comment]
> If I get a moment this week, I will try to look back at this. Sorry that you're still having issues @ricktiger. I know where the issue is, I've just not had a moment to patch it yet.

The current master branch seems to still have this bug. If you could point me in the right direction, I can probably fix the issue?

**Comment by coustack** [comment]
I was testing the changes and now receive the error of "json node is missing child properties" when creating a new contact. Any ideas? The JSON in the developer portal is:
Post: /contacts/v1/contact
Request: [{"property":"email","value":"johnSmith@person.com"}]

**Comment by JesseKlaasse** [comment]
Eventually I gave up, and I have created an own package based on HubSpot.NET, but then for the v3 version of the API, and with async/await support: https://github.com/IndiciaConnectivity/Indicia.HubSpot/

**Comment by bburrell** [comment]
> Eventually I gave up, and I have created an own package based on HubSpot.NET, but then for the v3 version of the API, and with async/await support: https://github.com/IndiciaConnectivity/Indicia.HubSpot/

@JesseKlaasse where you thinking of putting a license on your Indicia.Hubspot project? I would be interested in using it. Thanks.

**Comment by JesseKlaasse** [comment]
@bburrell I have added a MIT license, so please go ahead and use it.

**Comment by bburrell** [comment]
Thanks @JesseKlaasse  that is awesome!

**Comment by VladMstv** [comment]
@JesseKlaasse I'm currently using your library and I miss the ability to use the constructor of HubSpotApi like in HubSpot.NET or with some method of IHubSpotApi to authorize with ApiKey or OAuth credentials on the fly, not on startup. I will be really grateful if you consider adding it to your library. Thank you!
